cloudify_aws/s3/bucket.py

```diff
@@ -0,0 +1,178 @@
+########
+# Copyright (c) 2015 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+# Stdlib imports
+import string
+
+# Third party imports
+from botocore.exceptions import ClientError
+
+# Cloudify imports
+from cloudify import ctx
+from cloudify.decorators import operation
+from cloudify.exceptions import NonRecoverableError
+
+# This package imports
+from cloudify_aws import constants
+from cloudify_aws.base import AwsBaseNode
+from cloudify_aws.connection import EC2ConnectionClient
+
+
+@operation
+def create(args=None, **_):
+    return Bucket().created(args)
+
+
+@operation
+def delete(args=None, **_):
+    return Bucket().deleted(args)
+
+
+class Bucket(AwsBaseNode):
+    def __init__(self, client=None):
+        client = client or EC2ConnectionClient().client3('s3')
+        super(Bucket, self).__init__(
+            constants.S3_BUCKET['AWS_RESOURCE_TYPE'],
+            constants.S3_BUCKET['REQUIRED_PROPERTIES'],
+            client=client,
+            resource_id_key=constants.S3_BUCKET['RESOURCE_ID_KEY'],
+        )
+        self.not_found_error = constants.S3_BUCKET['NOT_FOUND_ERROR']
+        self.get_all_handler = {
+            'function': self.client.list_buckets,
+            'argument': 'Bucket',
+            'results_key': 'Buckets',
+            'id_key': 'Name',
+        }
+        self.resource_id = ctx.node.properties['name']
+        if self.is_external_resource:
+            ctx.instance.runtime_properties[constants.EXTERNAL_RESOURCE_ID] = (
+                ctx.node.properties['name']
+            )
+
+    def create(self, args=None):
+        self.validate_bucket_name(ctx.node.properties['name'])
+
+        self.execute(
+            self.client.create_bucket,
+            {
+                'Bucket': ctx.node.properties['name'],
+                'ACL': ctx.node.properties['permissions'],
+            }
```

Comment:
I think in this case the abstraction makes more sense. It'll only be even slightly jarring to someone who is used to the AWS API, and since we're creating an abstraction layer that's not something I'd say we're generally expecting.

In any other cases, having to set the property of 'bucket' on a bucket to set its name feels really peculiar.

I could change permissions -> ACL, but I think for most people, permissions is a bit clearer. Still, let me know if you'd prefer it changed.